fabtests/pytest/efa/efa_common.py

```diff
@@ -122,3 +128,79 @@ def get_efa_device_names(server_id):
         assert itemlist[1] in ["rdm", "dgrm"]
         device_names.add(itemlist[0])
     return list(device_names)
+
+
+class EFAMultinodeClientServerTest:
+    """
+    This class establishes tests involving a single server process
+    and multiple client processes which can be initiated on multiple
+    nodes (configured via client_host_name_list).
+    """
+
+    def __init__(self, cmdline_args, base_command, client_hostname_list):
+        self._cmdline_args = cmdline_args
+        self._base_command = base_command
+        self._timeout = self._cmdline_args.timeout
+        self.client_hostname_list = client_hostname_list
+        self.numclient = len(client_hostname_list)
+        self._server_command = cmdline_args.populate_command(self._base_command + " -C {}".format(self.numclient), "server", self._timeout)
+        assert(isinstance(client_hostname_list, list) and len(client_hostname_list) >= 1)
+        self._client_command_list = []
+        for client_hostname in client_hostname_list:
+            cmdline_args_copy = copy.copy(cmdline_args)
+            cmdline_args_copy.client_id = client_hostname
+            self._client_command_list.append(cmdline_args_copy.populate_command(self._base_command, "client", self._timeout))
+
+
+    def run(self):
+        if self._cmdline_args.is_test_excluded(self._base_command):
+            pytest.skip("excluded")
+
+        server_outfile = NamedTemporaryFile(prefix="fabtests_server.out.").name
+
+        # start running
+        server_process = Popen(self._server_command + "> " + server_outfile + " 2>&1", shell=True)
+        sleep(1)
```

Comment:
For robustness, the server startup could be slow and we should allow the client to retry if it fails to connect to the server similar to [this](https://github.com/ofiwg/libfabric/blob/main/fabtests/pytest/common.py#L303)

Maybe we should consider inheriting the current ClientServerTest, e.g. `class MultiClientServerTest(ClientServerTest)`. WDYT?

Reply:
Done. See latest revision.